core/src/avm1/object/array_object.rs

```diff
@@ -0,0 +1,281 @@
+use crate::avm1::property::Attribute;
+use crate::avm1::{Activation, Error, Object, ObjectPtr, ScriptObject, TObject, Value};
+use gc_arena::{Collect, GcCell, MutationContext};
+use std::borrow::Cow;
+use std::fmt;
+
+#[derive(Clone, Copy, Collect)]
+#[collect(no_drop)]
+pub struct ArrayObject<'gc>(GcCell<'gc, ScriptObject<'gc>>);
+
+impl fmt::Debug for ArrayObject<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("ArrayObject").finish()
+    }
+}
+
+impl<'gc> ArrayObject<'gc> {
+    pub fn empty(gc_context: MutationContext<'gc, '_>, proto: Option<Object<'gc>>) -> Self {
+        Self::new(gc_context, proto, &[])
+    }
+
+    pub fn new(
+        gc_context: MutationContext<'gc, '_>,
+        proto: Option<Object<'gc>>,
+        elements: &[Value<'gc>],
+    ) -> Self {
+        let base = ScriptObject::object(gc_context, proto);
+        for (i, &value) in elements.iter().enumerate() {
+            base.define_value(gc_context, &i.to_string(), value, Attribute::empty());
+        }
```

Comment:
Applied your first suggestion.
But unfortunately, there's a borrowing problem with the second one.
For example, consider `array_init_array` in `core/src/avm1/activation.rs`:
```rs
ArrayObject::new(
    self,
    (0..num_elements as i32).map(|_| self.context.avm1.pop()),
)
```
because `self` is `&mut Activation`, it cannot be borrowed to both `ArrayObject::new` and the closure.
I think the appropriate solution would be to somehow redesign the `Activation`, `UpdateContext` etc. layout.